rust/Earthfile

```diff
@@ -34,21 +41,59 @@ CARGO:
 REMOVE_SOURCE_FINGERPRINTS:
     COMMAND
-    COPY +get-stoml/stoml /tmp/stoml
-    ARG source_libs=$(find . -name Cargo.toml -exec bash -c '/tmp/stoml {} package.name; printf "\n"' \\;)
-    RUN --mount=type=cache,mode=0777,sharing=shared,target=target \
-      find target -name .fingerprint; \
-      fingerprint_folders=$(find target -name .fingerprint) ; \
-      for fingerprint_folder in $fingerprint_folders; do \
-        cd $fingerprint_folder; \
-        for source_lib in $source_libs; do \
-          find . -maxdepth 1 -regex "\./$source_lib-[^-]+" -exec rm -rf {} \; ;\
-        done \
-      done
+    COPY +get-tomljson/tomljson /tmp/tomljson
+    COPY +get-jq/jq /tmp/jq
+    DO +RUN_WITH_CACHE --command="set -e;
+      source_libs=\$(find . -name Cargo.toml -exec bash -c '/tmp/tomljson {} | jq -r .package.name; printf \"\\n\"' \\;) ;
+      fingerprint_folders=\$(find target -name .fingerprint) ;
+      echo \"deleting fingerprints:\";
+      for fingerprint_folder in \$fingerprint_folders; do
+        cd \$fingerprint_folder;
+        for source_lib in \$source_libs; do
+          find . -maxdepth 1 -regex \"\./\$source_lib-[^-]+\" -exec bash -c 'readlink -f {}; rm -rf {}' \; ;
+        done
+      done"
+
+# get-tomljson gets the portable tomljson binary
+get-tomljson:
+    FROM alpine:3.18.3
+    RUN wget -O tomljson.tar.xz https://github.com/pelletier/go-toml/releases/download/v2.1.0/tomljson_2.1.0_linux_amd64.tar.xz; \
+        tar -xf tomljson.tar.xz; \
+        chmod +x tomljson
+    SAVE ARTIFACT ./tomljson tomljson
 
-# get-stoml gets the portable stoml binary
-get-stoml:
+# get-jq gets the portable jq binary
+get-jq:
     FROM alpine:3.18.3
-    RUN wget -O stoml https://github.com/freshautomations/stoml/releases/download/v0.7.1/stoml_linux_amd64; \
-       chmod +x stoml
-    SAVE ARTIFACT ./stoml stoml
+    RUN apk add jq;
+    ARG jq=$(which jq)
+    SAVE ARTIFACT $jq jq
+
+# INSTALL_CARGO_SWEEP installs cargo-sweep if it doesn't exist already
+INSTALL_CARGO_SWEEP:
+    COMMAND
+    DO +RUN_WITH_CACHE --command="set -e;
+        if [ ! -f \$CARGO_HOME/bin/cargo-sweep ]; then
+          cargo install cargo-sweep;
+        fi;"
+
+# RUN_WITH_CACHE runs the passed command with the CARGO caches mounted.
+# Arguments:
+#   - command (required): Command to run, can be any expression.
+#
+# This implementation is not expected to significantly change. Prefer using the
+# `CARGO` UDC if you can, so you can get future improvements transparently.
+RUN_WITH_CACHE:
+    COMMAND
+    ARG command
+    ARG ORIGINAL_CARGO_HOME=$CARGO_HOME
+    ENV CARGO_INSTALL_ROOT=$CARGO_HOME
+    ENV CARGO_HOME="/earthly/.cargo"
+    RUN --mount=type=cache,mode=0777,id="earthly-cargo-cache",sharing=shared,target=/earthly \
+        --mount=type=cache,mode=0777,target=target \
+    set -e; \
+    mkdir -p $CARGO_HOME; \
+    mkdir -p /earthly/bin ;\
```

Comment:
I think you've removed all the uses of `/earthly/bin/` by now?

Reply:
cargo itself is in that folder

Reply:
Sorry, I was getting muddled between `$CARGO_HOME/bin` AKA `/earthly/.cargo/bin` and `/earthly/bin`. But after untangling that aspect I'm still unsure what is going on here...

What is the user of `/earthly/bin`? I think `cargo` is not there, or at least I can't see how it got there if it is. In the images I use in my `Earthfiles` it is in

```console
$ docker run -it rust:1.73.0-bookworm which cargo
/usr/local/cargo/bin/cargo
```
and I suppose it is image specific.

If cargo was in `/earthly/bin/cargo` then it would have been shadowed by the mount, wouldn't it, and then `mkdir` cannot save you?

Reply:
Fair enough :) 
- `/earthly/bin` was just an orphan reference I've just removed
- `$ORIGINAL_CARGO_HOME` is the cargo home in the target filesystem. It contains the cargo and rust binaries and the installed crates. We are only interested in its `bin` folder, that is on the `$PATH`
- `$CARGO_HOME` (AKA /earthly/.cargo/bin) is the new home created by the UDC. We will be using all their subfolders but the `bin` one, that will never be used (see https://github.com/earthly/lib/pull/20#issuecomment-1790620515)

Reply:
Added some comments in the source code. Hope is clearer now